### DISCUSSION TITLE
Compute Crystal Plasticity Thermal Eigenstrain

### DISCUSSION THREAD

**Discussion by Emmanuel-324** [idaholab/moose#29188] [Q&A Modules: Solid mechanics]
@sapitts I just had a couple of questions about the compute crystal plasticity thermal eigenstrain.

First is the temperature in Kelvin or degrees Celsius (I have assumed it's Kelvin)

Why is the stress output in the ''multiple_eigenstrains_test_out.csv'' negative(compressive)? My initial thought is that the domain wants to expand due to the thermal expansion coefficient being applied, and there are restricted boundaries that may cause compressive forces. However, the top and right boundaries are free of displacement and allow deformation, so I am kind of confused.

Lastly, why is the eigenstrain_deformation_gradient more than one isn't that more than 100% eigen strain contribution in the system?

I checked the referenced papers, but could not get any substantial evidence.

Any help?

**Comment by sapitts** [comment]
Hi @Emmanuel-324,

You are correct, the default temperature units are Kelvin.

The mesh in the `multiple_eigenstrains_test.i` is constrained on both the front and back in the z-direction, and so the compressive stress (and strain) will come from these boundary conditions. In this case the eigenstrains introduce a 3D volume change in the lattice, and the crystal must accommodate this increase in volume due to the eigenstrain with a comparative decrease in the crystal volume: this crystal volume change corresponds to a negative strain and a negative stress.

For the deformation gradient, recall the definition of the deformation gradient: the sum of the identity tensor and the partial derivative of the displacement with respect to the reference frame:
$$\mathbf{F} = \mathbf{I} + \frac{\partial \mathbf{u}}{\partial \mathbf{X}} $$
With this definition in mind, we know that a deformation gradient diagonal component (i.e. $F_{ii}$) in the absence of displacement will be equal to 1 (from the identity tensor). Tensile displacement is reflected as a deformation gradient value greater than 1.0; compressive displacement is reflected as a deformation gradient value less than 1.0.

See this website for a review of the deformation gradient definition and a few 2D examples of calculating the deformation gradient: https://www.continuummechanics.org/deformationgradient.html

**Reply by Emmanuel-324** [reply]
Thanks for the explanation and the resource.
Below are some relevant snippets of my scrip
```
[Mesh]
  [file]
     type = FileMeshGenerator
     file = Conc1_out.e-s123
     use_for_exodus_restart = true
   []
 []

[Variables]
   [./disp_x]
   [../]
   [./disp_y]
   [../]
 
   # order parameter 0
   [./eta0]
     initial_from_file_var = eta1
   [../]
   # order parameter 1
   [./eta1]
     initial_from_file_var = eta3
   [../]
 []
[Functions]
  [temperature_function]
    type = ParsedFunction
    value = 'if(t < 0.002, 300 * t / 0.002, 300)'
  []
 [] 
 [BCs]
  [symmy]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0
  []
  
  [symmx]
    type = DirichletBC
    variable = disp_x
    boundary = left
    value = 0
  []
  [tdisp]
    type = FunctionDirichletBC
    variable = disp_x
    boundary = right
    function = 0.1*t
  []
 []
 [eigenstrain_0]
    type = ComputeCrystalPlasticityThermalEigenstrain
    eigenstrain_name = eigenstrain_0
    deformation_gradient_name = deformation_gradient_1
    temperature = temperature
    thermal_expansion_coefficients = '1.2e-05 2e-05 3e-05'
    base_name = phase0
  []

[Kernels]
   [./eta0_dt]
     type = TimeDerivative
     variable = eta0
   [../]
   [./eta1_dt]
     type = TimeDerivative
     variable = eta1
   [../]
   [./TensorMechanics]
      displacements = 'disp_x disp_y'
      add_variables = true
      generate_output = stress_xx
      strain = FINITE
   []
 []
```

I have this crystal plasticity script above that has some of the results below:

![image](https://github.com/user-attachments/assets/26f828cd-2055-4fe1-b54a-4729d5bde8a2)
![image](https://github.com/user-attachments/assets/1c47d66c-eb2c-4bbd-a2b2-88d430be3631)
![image](https://github.com/user-attachments/assets/5a1fae7c-7c8e-4be4-bca9-7ab98edaaefe)
![image](https://github.com/user-attachments/assets/e98138a0-902a-488b-8632-30cd145ee95e)

I kept the temperature constant because that was not the focus. I only need the eigen contribution.

My first question is to confirm whether how I calculated my eingenstrain for the plots above is correct. I did it as show in the picture below: 
![image](https://github.com/user-attachments/assets/88cd6dfe-587f-4e8d-a0ac-a95f263a3761).

My second question: Yes my assumption for the compressive stress was because of the boundary conditions as you said. However, after applying displacement in the x direction of my script even though the temperature was constant the graph(Stress vs time) sees the plot moving from compressive to tensile stress—any reason why would be helpful. 

Lastly, when I run my script without displacement the computational time is minimal (it takes about 2 hours to get to 100s), but when I add the displacement factor the same script takes about 3 days to get to a timestep of 12.6s. Is there a way I can improve on this?

**Reply by sapitts** [reply]
is the "displacement factor" the constant eigenstrain?

**Reply by Emmanuel-324** [reply]
The constant eigenstrain is due to the constant temperature and the same expansion coefficient in both x and y. The displacement factor is for the plastic deformation in the crystal plasticity model.
